src/models/whisper.cpp

```diff
@@ -66,21 +256,174 @@ RoamingArray<float> Whisper_State::Run(int current_length, RoamingArray<int32_t>
+        input_names_.push_back("cache_indirection");
+        inputs_.push_back(cache_indirection_.get());
+
+        auto data = gpu_span<int32_t>{cache_indirection_->GetTensorMutableData<int32_t>(),
+                                      static_cast<size_t>(params_->batch_size) * params_->search.num_beams * params_->search.max_length};
+        cudaMemsetAsync(data.data(), 0, data.size_bytes(), params_->cuda_stream);
+#endif
+      }
+
+      if (model_.session_info_->HasOutput("output_cross_qk_0")) {
+        auto layer_count = model_.config_->model.decoder.num_hidden_layers;
+        auto type = model_.session_info_->GetOutputDataType("output_cross_qk_0");
+        std::array<int64_t, 4> shape{params_->BatchBeamSize(), model_.config_->model.decoder.num_attention_heads, 1, 1500};
+        for (int i = 0; i < layer_count; i++) {
+          char string[64];
+          snprintf(string, std::size(string), "output_cross_qk_%d", i);
+          output_cross_qk_names_.emplace_back(string);
+          output_cross_qk_.emplace_back(OrtValue::CreateTensor(*model_.allocator_device_, shape, type));
+
+          output_names_.emplace_back(output_cross_qk_names_.back().c_str());
+          outputs_.emplace_back(output_cross_qk_.back().get());
+        }
+      }
+
+      {
+        bool search_buffers = false;
+        UpdateInputsOutputs(next_tokens, next_indices, current_length, search_buffers);
+      }
       break;
+
+    case RunState::Decoder: {
+      bool search_buffers = true;
+      UpdateInputsOutputs(next_tokens, next_indices, current_length, search_buffers);
+    } break;
   }
 
   State::Run(*model_.session_decoder_, *model_.run_options_, batch_size);
   return logits_.Get();
 }
 
-void Whisper_State::UpdateInputsOutputs(const RoamingArray<int32_t>& next_tokens, RoamingArray<int32_t> beam_indices, int current_length) {
+void Whisper_State::UpdateInputsOutputs(const RoamingArray<int32_t>& next_tokens, RoamingArray<int32_t> beam_indices, int current_length, bool search_buffers) {
   decoder_input_ids_.Update(next_tokens);
   kv_cache_.Update(beam_indices.GetCPU(), current_length);
   logits_.Update();
+
+  if (past_sequence_length_) {
+    auto data = past_sequence_length_->GetTensorMutableData<int32_t>();
+    *data = current_length - 1;
+  }
+
+  if (!search_buffers) {
+    // No need to update cache indirection and cross QK search buffers
+    // when preparing to run decoder for the first time.
+    if (cache_indirection_)
+      inputs_[cache_indirection_index_] = cache_indirection_.get();
+    return;
+  }
+
+  if (cache_indirection_) {
+#if USE_CUDA
+    gpu_span<int32_t> beam_indices_gpu = beam_indices.GetGPU();
+    cuda_unique_ptr<int32_t> beam_indices_ptr;
+    if (beam_indices_gpu.empty()) {
+      beam_indices_ptr = CudaMallocArray<int32_t>(params_->batch_size, &beam_indices_gpu);
+      std::vector<int32_t> beam_indices_cpu(params_->batch_size, 0);
+      std::iota(beam_indices_cpu.begin(), beam_indices_cpu.end(), 0);
```

Comment:
Should we do this elsewhere instead? This initialization isn't Whisper specific.

Reply:
I think this should be local to file since the beam indices (for non-beam-search search algorithms) is only needed by this whisper model due to DMMHA. Other models will probably never need to care about beam_indices in case beam-search is not used. 
If you think we can move the logic to some other place, I can try to move it.